### ISSUE TITLE
Enhancement: [no-redundant-type-constituents] Option to ignore certain types

### ISSUE THREAD

**Issue by oliveryasuna** [typescript-eslint/typescript-eslint#10036]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-redundant-type-constituents/

### Description

It is often useful to include redundant type constituents for both IDE intelligence and documentation.

For example:

```typescript
type MyType = unknown | undefined;
```

The above informs the IDE and documentation readers that, while the type is unknown, it is expected that it could be undefined. This is informative in the same way that a type without `unknown` would be, e.g., `Foo | undefined`.

Now, we get to the weird part... Should the new option specify ignored types. E.g., `"ignoredTypes": ["undefined"]`, would make `unknown | undefined` valid because it would strip `undefined`. Or, should it accept types for which the rule is disabled, e.g., `"ignoredTypes": ["unknown"]`, would make `unknown | undefined` valid because `unknown` is part of the union? I would like to open a discussion with the community.

The documentation actually shows an example of when "this rule might not be for you": https://typescript-eslint.io/rules/no-redundant-type-constituents/#when-not-to-use-it. It's not exactly the same, but still relevant. The point is, there are use cases for modifying the behavior or this rule.

### Fail

```TypeScript
See description.
```

### Pass

```TypeScript
See description.
```

**Comment by bradzacher** [comment]
> It is often useful to include redundant type constituents for both IDE intelligence

There is no IDE intelligence that will come from redundant types. The IDE features are powered by TS and it will elide the redundant types - which is really the point of the rule.

---

> while the type is unknown, it is expected that it could be undefined. This is informative in the same way that a type without unknown would be, e.g., Foo | undefined.

I'm not sure I understand this argument.

`unknown` includes `undefined` and the only way to refine out of `unknown` is to use conditions to guard against both `null` and `undefined`.

I don't see how it is comparable to `Foo | undefined`.
`Foo` by itself would not accept `undefined`, nor would it require any guards against `undefined` when using it, and `Foo | undefined` declares the `undefined` which allows the assignment and enforces that you handle it.
On the contrary - `unknown` by itself already fits the latter description - it allows assignment of `undefined` and enforces that you handle it. Adding `| undefined` does not add any features.

**Comment by kirkwaiblinger** [comment]
I am similarly perplexed to @bradzacher here... 
Two thoughts to add, though
1. IIRC, Python does track redundant unions like `Any | Literal[0]`- maybe there is a useful lesson for us from the usage there? (Or maybe I'm making up this memory)
2. Do you have a habit of using `unknown` to mean "anything but except undefined/null"? If so, maybe `Nonnullable<unknown>` would be a better fit for those cases, and you reserve `unknown` for when it's _really, really_ anything?

**Comment by bradzacher** [comment]
It has been two weeks without response from OP.
Closing out.
